### PR TITLE
Create Kyma step

### PR DESCRIPTION
**Pull request by piotrmiskiewicz**
**Description**

Changes proposed in this pull request:
- implementation of a step which creates/updates Kyma resource

**Related issue(s)**

### CODE REVIEW

**Comment by piotrmiskiewicz** [comment]
Open questions:
 - what should be the name of the Kyma resource?
 - what should be the pattern of labels?
 - do we need any annotations?

File: components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	require.Zero(t, backoff)
+	aList := unstructured.UnstructuredList{}
+	aList.SetGroupVersionKind(schema.GroupVersionKind{Group: "operator.kyma-project.io", Version: "v1alpha1", Kind: "KymaList"})
+
+	operation.K8sClient.List(context.Background(), &aList)
+	assert.Equal(t, 1, len(aList.Items))
+	assertLabelsExists(t, aList.Items[0])
+}
+
+func TestUpdateingKymaResourceIfExists(t *testing.T) {
```
Comment (jaroslaw-pieszka) [inline components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step_test.go:63]:
```suggestion
func TestUpdatingKymaResourceIfExists(t *testing.T) {
```

**Comment by wozniakjan** [comment]
ptal @piotrmiskiewicz, `go fmt` in `pre-control-plane-components-kyma-environment-broker` job is complaining about formatting
```
[0100] | ✗ some files are not properly formatted or contain not formatted imports. To repair run make imports
[0100] | ./internal/process/provisioning/apply_kyma_step.go
```

**Review by wozniakjan — COMMENTED** [review]
ptal @piotrmiskiewicz, `go fmt` in `pre-control-plane-components-kyma-environment-broker` job is complaining about formatting
```
[0100] | ✗ some files are not properly formatted or contain not formatted imports. To repair run make imports
[0100] | ./internal/process/provisioning/apply_kyma_step.go
```

File: components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	existingKyma.SetGroupVersionKind(kymaGVK)
+	err = k8s.Get(context.Background(), client.ObjectKey{
+		Namespace: template.GetNamespace(),
+		Name:      template.GetName(),
+	}, &existingKyma)
+	switch {
+	case err == nil:
+		logger.Infof("Kyma resource already exists, updating")
+		a.addLabelsAndName(operation, &existingKyma)
+		err = k8s.Update(context.Background(), &existingKyma)
```
Comment (wozniakjan) [inline components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go:65]: nit: small performance optimization, I am considering doing diff between `template` and `existingKyma` in https://github.com/kyma-project/control-plane/pull/2216 (not there yet), we can prevent API call if there is no diff..

File: components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+func (a *ApplyKymaStep) Name() string {
+	return "Apply_Kyma"
+}
+
+func (a *ApplyKymaStep) Run(operation internal.Operation, logger logrus.FieldLogger) (internal.Operation, time.Duration, error) {
+	if operation.K8sClient == nil {
+		return a.operationManager.OperationFailed(operation, "operation does not contain initialized k8s client", nil, logger)
+	}
+	k8s := operation.K8sClient
```
Comment (wozniakjan) [inline components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go:46]: isn't `operation.K8sClient` a client to SKR's kube-api? I thought the kyma resource would live in the same cluster as kyma control plane. At the moment https://github.com/kyma-project/control-plane/pull/2216 is implemented to save the kubeconfig to KCP so the lifecycle manager can get the kubeconfig to SKR.

File: components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go (hunk 2 above)
Comment (piotrmiskiewicz) [reply components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go:65]: KEB is not the only one component who manage this resource. The spec could be changed. Also somebody else can modify labels. I can imagine that we can check if labels where modified, then decide if we need the update, but it is a rare case in my opinion

Comment (wozniakjan) [reply components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go:65]: I think we agree, KEB doesn't need to call `Update` unless there is a diff between `template` and `existingKyma`. But can be considered an optimization candidate for the future.

File: resources/kcp/charts/kyma-environment-broker/templates/deployment.yaml [hunk 4]
```diff
@@ -83,6 +83,9 @@ spec:
               value: "{{ .Values.broker.operationTimeout }}"
             - name: APP_RECONCILER_URL
               value: "{{ .Values.reconciler.URL }}"
+              LifecycleManagerIntegrationDisabled
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply resources/kcp/charts/kyma-environment-broker/templates/deployment.yaml:86]: it was by accident

**Review by wozniakjan — APPROVED** [review]
/lgtm
/approve

File: components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return operation, 0, nil
+}
+
+func (a *ApplyKymaStep) addLabelsAndName(operation internal.Operation, obj *unstructured.Unstructured) bool {
+	oldLabels := obj.GetLabels()
+	steps.ApplyLabelsForLM(obj, operation)
+	obj.SetName(steps.KymaName(operation))
+	if !reflect.DeepEqual(obj.GetLabels(), oldLabels) {
```
Comment (jaroslaw-pieszka) [inline components/kyma-environment-broker/internal/process/provisioning/apply_kyma_step.go:87]: `return !reflect.DeepEqual(obj.GetLabels(), oldLabels)` ?

**Review by jaroslaw-pieszka — APPROVED** [review]

**Pull request merged**